### PR TITLE
GTEST: Add ucp_test::mapped_buffer helper class

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: test/gtest/ucp/test_ucp_fence.cc [hunk 1]
```diff
@@ -117,71 +117,23 @@ class test_ucp_fence : public test_ucp_atomic {
...
-        params.flags      = GetParam().variant;
-        if (params.flags & UCP_MEM_MAP_FIXED) {
-            params.address  = ucs::mmap_fixed_address();
-            params.flags   |= UCP_MEM_MAP_ALLOCATE;
-        } else {
-            memheap = malloc(memheap_size);
-            params.address = memheap;
-            params.flags = params.flags & (~UCP_MEM_MAP_ALLOCATE);
-        }
+        mapped_buffer buffer(memheap_size, receiver(), 0);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_fence.cc]: @hoopoepg variant should not really be used here: it was supposed to be MAP_NONBLOCK from test_ucp_memheap, but atomic test overrode enum_test_params() with atomic mode value. so we don't really need to allocate memory with UCP, and can let mem_buffer allocate it.

File: test/gtest/common/mem_buffer.h [hunk 2]
```diff
@@ -59,9 +59,8 @@ class mem_buffer {
     /* return the string name of a memory type */
     static std::string mem_type_name(ucs_memory_type_t mem_type);
 
-    mem_buffer(void *ptr, size_t size, ucs_memory_type_t mem_type);
```
Comment (yosefe) [inline test/gtest/common/mem_buffer.h]: removed unimplemented ctor

File: test/gtest/common/mem_buffer.h [hunk 3]
```diff
@@ -74,9 +73,9 @@ class mem_buffer {
 
     static uint64_t pat(uint64_t prev);
 
-    ucs_memory_type_t m_mem_type;
-    void              *m_ptr;
-    size_t            m_size;
```
Comment (yosefe) [inline test/gtest/common/mem_buffer.h]: make mem_buffer class immutable

File: test/gtest/ucp/ucp_test.cc [hunk 4]
```diff
@@ -729,3 +729,52 @@ void ucp_test_base::entity::ep_destructor(ucp_ep_h ep, entity *e)
...
+    EXPECT_UCS_OK(status);
+}
+
+ucs::handle<ucp_rkey_h> ucp_test::mapped_buffer::rkey(const entity& entity) const
+{
+    ucp_rkey_h rkey;
+
+    ucs_status_t status = ucp_ep_rkey_unpack(entity.ep(), m_rkey_buffer, &rkey);
+    ASSERT_UCS_OK(status);
+    return ucs::handle<ucp_rkey_h>(rkey, ucp_rkey_destroy);
```
Comment (yosefe) [inline test/gtest/ucp/ucp_test.cc]: ucs::handle operator= can take over object ownership

File: test/gtest/ucp/test_ucp_fence.cc [hunk 5]
```diff
@@ -117,71 +117,23 @@ class test_ucp_fence : public test_ucp_atomic {
...
-        ASSERT_UCS_OK(status);
-
-        status = ucp_ep_rkey_unpack(sender().ep(), rkey_buffer, &rkey);
-        ASSERT_UCS_OK(status);
-
-        ucp_rkey_buffer_release(rkey_buffer);
-
-        run_workers(send1, send2, &sender(), rkey, memheap, 1, &error);
+        run_workers(send1, send2, &sender(), buffer.rkey(sender()),
+                    buffer.ptr(), 1, &error);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_fence.cc]: temp object ucs::handle> is created, and destroyed after the function returns

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 25 workers (click for details)</summary>

**Note:** the logs will be deleted after 23-Nov-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/52faac4d042e8080751fa90a61d6612da13bd0fd/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/4565dbad8726bd3b06a274f10ac717c49dd3af4c/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/f1f07a82cca4e3a260a8173478fcb400ae1576c8/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/bc69ec52d3fc77c642880834f8a689730bfd2c9a/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/4e22558eabb9b2913cef2a5358019e08106b992f/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/74328fece5c549892664c01af45c7ef8bdd26d45/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/6d30f3dd21a80fd3b83d95cd6733fc90172e4129/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/67d8340590e5e26230a9dab4291935d6b297b2b3/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/b258693ae6faba8ff775b41b7bd5273a77a9d212/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W0](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/b2035d6a10b92561fa3a7856fb53db1c9713fe39/hpc-test-node-gpu_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W1](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/2a801b690e096fe6105b66ef8c01dc65356c77ec/hpc-test-node-gpu_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W2](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/ba876a2c4fc0425eb8a0a5e0256dae3f462f4752/hpc-test-node-gpu_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-gpu_W3](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/09c3dcf61af46f84f920dce29fad737d9c5f1a7e/hpc-test-node-gpu_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W0](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/0719fdd052c30afda5c1bb15c63278a9ada4a123/hpc-test-node-legacy_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W1](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/90473a4d949610e71b17ecf0d84ce1f9fc5b2e0e/hpc-test-node-legacy_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W2](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/0c2e18caf9466551ae254e59accfe22d3c4bdd32/hpc-test-node-legacy_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-legacy_W3](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/977c069ae158f43ea1a339d7e5850e1c98129be1/hpc-test-node-legacy_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W0](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/20651b17ad4bc99dcd0c65d050221a3af538927e/hpc-test-node-new_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W1](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/9599fb42984c87cc913e1eed6d9c8c78a8deef6a/hpc-test-node-new_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W2](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/7f41567441feab09e89a00a9e445cb8e46d8a451/hpc-test-node-new_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-new_W3](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/659d35b3d5f88d29983f3d7cf73a94657e51d838/hpc-test-node-new_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/82988e13150b4fab3659356cc3d28cce72bb3685/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/28c77f5a73b33b4d770941d6e26d32852189bb6b/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/14630f80ee5fb33e0217d7534c6bd22d0f49321c/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/10810e0c73cfb29fa1fc5e8244dfd9ec/raw/beaf01c50a4f751a5417e4268aa6ed86de64d6bf/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by yosefe** [comment]
failure is #4419 
bot:pipe:retest

File: test/gtest/ucp/ucp_test.cc (hunk 4 above)
Comment (hoopoepg) [reply test/gtest/ucp/ucp_test.cc]: there is really hacky magic in ucs::handle - it destroys original object (like uniq_ptr) which may be surprise for others :)

but ok, see your point

Comment (yosefe) [reply test/gtest/ucp/ucp_test.cc]: i was too lazy to implement real shared_ptr :)

**Pull request merged**